src/main/kotlin/org/rust/ide/refactoring/move/RsMoveFilesOrDirectoriesHandler.kt

```diff
@@ -0,0 +1,85 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.refactoring.move
+
+import com.intellij.lang.Language
+import com.intellij.openapi.actionSystem.DataContext
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiReference
+import com.intellij.refactoring.move.MoveCallback
+import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesHandler
+import com.intellij.refactoring.move.moveFilesOrDirectories.MoveFilesOrDirectoriesUtil
+import com.intellij.refactoring.util.CommonRefactoringUtil
+import org.rust.ide.experiments.RsExperiments
+import org.rust.lang.RsLanguage
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.ext.isEdition2018
+import org.rust.lang.core.psi.hasChildModules
+import org.rust.openapiext.isFeatureEnabled
+
+class RsMoveFilesOrDirectoriesHandler : MoveFilesOrDirectoriesHandler() {
+
+    override fun supportsLanguage(language: Language): Boolean = language.`is`(RsLanguage)
+
+    override fun adjustTargetForMove(dataContext: DataContext?, targetContainer: PsiElement?): PsiElement? {
+        return (targetContainer as? RsFile)?.containingDirectory ?: targetContainer
+    }
+
+    private fun PsiElement.canMove(): Boolean {
+        if (this !is RsFile) return false
+
+        // TODO: support move files with its child modules
+        if (ownsDirectory) return false
+        if (hasChildModules()) return false
+
+        // TODO: support path attribute on mod declaration
+        if (pathAttribute != null) return false
+
+        return modName != null
+            && crateRoot != null
+            && crateRelativePath != null
+            && !isCrateRoot
+            && isEdition2018  // TODO: support 2015 edition
+    }
+
+    override fun canMove(elements: Array<out PsiElement>, targetContainer: PsiElement?, reference: PsiReference?): Boolean {
+        if (!isFeatureEnabled(RsExperiments.MOVE_REFACTORING)) return false
+        if (!elements.all { it.canMove() }) return false
+
+        // TODO: support move multiply files
+        if (elements.size > 1) return false
+
+        val adjustedTargetContainer = adjustTargetForMove(null, targetContainer)
+        return super.canMove(elements, adjustedTargetContainer, reference)
```

Comment:
Looks like it always returns `false` and doesn't use `adjustedTargetContainer` anyhow

Reply:
It is used in [`MoveFilesOrDirectoriesHandler#isValidTarget`](https://github.com/JetBrains/intellij-community/blob/664d0fc8c79325456f8a1d492240695705bf2af0/platform/lang-impl/src/com/intellij/refactoring/move/moveFilesOrDirectories/MoveFilesOrDirectoriesHandler.java#L66). There is complex call chain:

* Our code calls [`MoveFilesOrDirectoriesHandler#canMove`](https://github.com/JetBrains/intellij-community/blob/664d0fc8c79325456f8a1d492240695705bf2af0/platform/lang-impl/src/com/intellij/refactoring/move/moveFilesOrDirectories/MoveFilesOrDirectoriesHandler.java#L58)
* It calls [`MoveHandlerDelegate#canMove`](https://github.com/JetBrains/intellij-community/blob/78a91459c4d8454b2d5df36ade108807ef772373/platform/lang-impl/src/com/intellij/refactoring/move/MoveHandlerDelegate.java#L39)
* It calls another [`MoveHandlerDelegate#canMove`](https://github.com/JetBrains/intellij-community/blob/78a91459c4d8454b2d5df36ade108807ef772373/platform/lang-impl/src/com/intellij/refactoring/move/MoveHandlerDelegate.java#L47)
* It calls [`MoveFilesOrDirectoriesHandler#isValidTarget`](https://github.com/JetBrains/intellij-community/blob/664d0fc8c79325456f8a1d492240695705bf2af0/platform/lang-impl/src/com/intellij/refactoring/move/moveFilesOrDirectories/MoveFilesOrDirectoriesHandler.java#L63)

Reply:
Yes, I was wrong, thank you!